### PR TITLE
diff_package_list in Rex::Pkg::Redhat, for multiple pks with same name

### LINKED ISSUE #1149: pkg on_change runs even when package state did not change.
Target = centos 7 with all updates installed.

Rexfile:

```perl
use strict; use warnings;
use Rex -feature => [qw[1.3]];

user 'root';

task 'ttt' => sub {
   pkg 'postgresql-server', ensure => 'present',
       on_change => sub { say 'I should only run once!' };
};

```

Running 
```
rex -H target ttt
```
multiple times shows that on_change is invoked every time.

### PR DESCRIPTION
**Pull request by akarelas**
This PR is a proposal to fix #1149 by creating a `diff_package_list` method for the `Rex::Pkg::Redhat` module.

Please review and merge, or let me know how to improve it further.

## Checklist

- [x] based on top of latest source code 
- [x] changelog entry included 
- [x] automated tests pass 
- [x] git history is clean 
- [x] git commit messages are [well-written](https://chris.beams.io/posts/git-commit/#seven-rules)

### CODE REVIEW

**Comment by ferki** [comment]
Thanks for contributing, @akarelas!

Skimming over the content, and running the full test suite locally indicates this is a great starting point. Let's iterate towards merging together about any remaining details. I'll leave first review notes separately soon.

**Review by ferki — CHANGES_REQUESTED** [review]
Apparently the GitHub workflow for full testing failed to trigger due to merge conflicts. It looks like the new code is based on top of an earlier version of the default branch, which had new commits since that point.

Due to this, I unticked the related review checklist items for now. We'll track those based on the actual state of PR and test results.

Please rebase your changes on top of the current state of the default branch, and (force) push. With the conflicts solved, I expect to get the automated full test suite results from the GitHub workflows.

Let me know if you need help with the rebase and pushing again. While technically GitHub allows me to do this on your PR branch,  that would overwrite your history there, so I prefer do it more as a last resort than first.

**Comment by akarelas** [comment]
I based it on the latest default branch now.

**Comment by akarelas** [comment]
Reworded by Test commit

**Review by ferki — CHANGES_REQUESTED** [review]
Thanks for rebasing!

Now the full workflow has finished, we are getting closer, and see that solely the extra tests failing. This is highly promising, as the functionality seems to pass everywhere without breaking anything we already know.

The extra tests are about the code itself instead of what it does: code convention, code formatting, minimum required perl syntax, pod syntax, etc. The currently established incremental improvement idea regarding code conventions is "each change should make the code base cleaner, or at least keep it the same". In other words, while we don't need to fix everything that is already present, we need to ensure all new changes are clean.

Here are the single test failure reported about increased number of perlcritic violations compared to the default branch: https://github.com/RexOps/Rex/actions/runs/10554972139/job/29237791411#step:12:17

```
# ControlStructures::ProhibitPostfixControls: Got 313 violation(s).  Expected no more than 309.
# References::ProhibitDoubleSigils: Got 84 violation(s).  Expected no more than 81.
# TooMuchCode::ProhibitDuplicateLiteral: Got 3449 violation(s).  Expected no more than 3428.
# Too many Perl::Critic violations...
# Got a total of 461. Expected no more than 19433.
```

Let's fix those perlcritic feedback by iterating on the commits. It's fine to send additional commits on top which we can collapse into the current ones later, or just edit the commits locally, and (force) push again until everything is green.

Please see the perlcritic docs for each policy for their description (`perlcritic --doc $policy_name`), and run perlcritic to get detailed line info about what it complains about and where (`perlcritic --single-policy $policy_name $filename`). You may need to install the perlcritic policies locally (see `dzil listdeps --develop`.)

Let me know how much help you may need with tracking down and fixing these perlcritic issues. I'm happy to provide feedback accordingly, for example some ideas:

- you iterate on this PR on your own until everything is green
- I propose how to fix each, optionally with a diff you may apply
- I fix it completely on my own later

**Comment by akarelas** [comment]
Finished with extra tests. Can you please recheck?

**Review by ferki — COMMENTED** [review]
Thanks for the patience, while I went through everything in a way to enable meaningful feedback. I carefully read the code, and played with different situations, or alternative approaches.

Overall, on top of a review, it seems we are already through the red-green (fail-pass, test-code) steps of a standard TDD cycle, and it's time to refactor (improve based on what we learned) too.

The key code change seems to be treating the version and release metadata of a package together as a concatenated string to identify unique instances of the package. On Debian the criteria seemed to be more like name+arch, not sure arch plays a role on Red Hat yet. Do you have any experience or info about that?

The chosen implementation may perform better too, though I did not made any benchmarks yet. Mainly due to the Base and Redhat implementations do different things, which makes them harder to compare. Let's focus on correctness first.

Apart from that, anything else I noted down was related to the test approach. Since I felt showing a proposed change would be easier instead of only describing it, I decided to push a proposed commit to this branch.

What I also learned is the related `on_change` failure only seems to happen when the new version of the package is listed before the old version of the package with the same name.

Let me know what you think, and whether would you like to change anything else.

ps.: I'm not yet sure about how we may be able to test the actual `on_change` hook instead of only the underlying `diff_package_list()` implementation :thinking: I'd like to think a bit more about that part, while everything else gets addressed.

**Comment by ferki** [comment]
Oh, ouch, yes, the test workflow currently fails on older Strawberry Perl versions due to having outdated Mozilla::CA preinstalled while actions-setup-perl switched to "secure by default" cpanm calls. Let's see if that gets resolved upstream, or it's better to apply a workaround downstream here (and then rebase this.)

Let's track that separately, just keep in mind here, as the rest still applies.

**Comment by akarelas** [comment]
I think dclone is needed, because the old diff_package_list subroutine (which we want to fail) edits the pck entries, and without dclone the test would pass on the old code. But let's check again (when I find some time).

File: t/package.t [hunk 1]
```diff
@@ -62,4 +65,75 @@ subtest 'local package installation' => sub {
   lives_ok { $pkg->update('test_package') }, 'update test package';
 };
 
+subtest 'redhat package list diffs' => sub {
+  ## no critic (ProhibitDuplicateLiteral)
```
[the remark this answers is not in the document]
Comment (ferki) [reply t/package.t]: I moved this `## no critic` note in my follow-up commit to apply only to test data.

File: t/package.t [hunk 2]
```diff
@@ -62,4 +65,75 @@ subtest 'local package installation' => sub {
...
+
+  plan tests => 4;
+
+  my $rh_pkg = Rex::Pkg::Redhat->new;
```
[the remark this answers is not in the document]
Comment (ferki) [reply t/package.t:69]: Let's keep it like this for now, and we may still refactor it later as needed.

File: t/package.t [hunk 3]
```diff
@@ -62,4 +65,75 @@ subtest 'local package installation' => sub {
...
+  );
+
+  @plist1 = @{ dclone( \@orig ) };
+  @plist2 = @{ dclone( \@orig ) };
+  shift @plist2;
+  @expected = ( { %{ $orig[0] }, action => 'removed' } );
+
+  @mods = $rh_pkg->diff_package_list( \@plist1, \@plist2 );
+  cmp_deeply( \@mods, \@expected,
+    'expected package modifications when only a kernel release is removed' );
```
[the remark this answers is not in the document]
Comment (ferki) [reply t/package.t]: My follow-up commit converts the new tests to a purely data-driven approach, and covers multiple modifications at once.

**Comment by ferki** [comment]
I decided to resolve the previous discussion points based on my follow-up commit about refining the related tests.

I also went ahead to rebase this on top of current default branch (and fix the merge conflict here in the changelog) to trigger a new test run.

I believe this is close to be merged, though let's see the fresh test results first :)

**Review by ferki — APPROVED** [review]

**Comment by ferki** [comment]
Thank you for contributing, @akarelas!
